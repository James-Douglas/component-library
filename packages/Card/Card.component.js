import React from 'react';
import PropTypes from 'prop-types';
import { spacingPropTypes } from '@comparethemarketau/manor-utils';
import { useId } from '@comparethemarketau/manor-hooks';
import StyledCard from './Card.styles';

const Card = ({
  id: propsId,
  children,
  margin,
  padding,
  className,
}) => {
  const id = useId(propsId);
  return (
    <StyledCard id={id} margin={margin} padding={padding} className={className}>
      {children}
    </StyledCard>
  );
};

Card.propTypes = {
  /**
   * Unique identifier for the card
   */
  id: PropTypes.string,
  /**
   * Card content
   */
  children: PropTypes.node,
  /**
   * Allows setting of margin values for the card
   */
  margin: spacingPropTypes,
  /**
   * Allows setting of padding values for the card
   */
  padding: spacingPropTypes,
  /**
   * Classes to be applied to the Card component
   */
  className: PropTypes.string,
  // eslint-disable-next-line react/forbid-prop-types
};

Card.defaultProps = {
  id: null,
  children: [],
  margin: ['4', '4', '16'],
  padding: ['16'],
  className: '',
};

export default Card;
