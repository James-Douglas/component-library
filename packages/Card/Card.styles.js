import styled from 'styled-components';
import { applySpacing } from '@comparethemarketau/manor-utils';

const StyledCard = styled.div`
  box-sizing: border-box;
  display: flex;
  width: 100%;
  min-width: ${({ theme }) => theme.spacing['80']};
  min-height: ${({ theme }) => theme.spacing['80']};
  background: ${({ theme }) => theme.card.background};
  box-shadow: ${({ theme }) => theme.card.shadow};
  border: ${({ theme }) => theme.card.border};
  ${({ margin, theme }) => applySpacing(theme, 'margin', margin)};
  ${({ padding, theme }) => applySpacing(theme, 'padding', padding)};
`;

export default StyledCard;
