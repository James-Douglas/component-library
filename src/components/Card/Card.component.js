import React from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider } from 'styled-components';
import getTheme from 'utils/getTheme';
import applySpacing, { spacingPropTypes } from 'utils/applySpacing';

const StyledCard = styled.div`
  display: flex;
  width: 100%;
  min-width: ${({ theme }) => theme.spacing['80']};
  min-height: ${({ theme }) => theme.spacing['80']};
  background: ${({ theme }) => theme.colors.white}; 
  box-shadow: ${({ theme }) => theme.boxShadow.progress}; 
  border: ${({ theme }) => `1px solid ${theme.colors.greyLight}`};   
  ${(props) => applySpacing('margin', props.margin)};
  ${(props) => applySpacing('padding', props.padding)}; 
`;

const Card = ({
  id, children, margin, padding,
}) => (
  <ThemeProvider theme={getTheme()}>
    <StyledCard id={id} margin={margin} padding={padding}>
      {children}
    </StyledCard>
  </ThemeProvider>
);

Card.propTypes = {
  /**
   * Unique identifier for the card
   */
  id: PropTypes.string.isRequired,
  /**
   * Card content
   */
  children: PropTypes.node,
  /**
   * Allows setting of margin values for the card
   */
  margin: spacingPropTypes,
  /**
   * Allows setting of padding values for the card
   */
  padding: spacingPropTypes,
};

Card.defaultProps = {
  children: [],
  margin: ['4', '4', '16'],
  padding: ['16'],
};

export default Card;
