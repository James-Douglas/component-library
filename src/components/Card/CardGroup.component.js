import React from 'react';
import PropTypes from 'prop-types';
import Container from '../Grid/Container/Container.component';
import Row from '../Grid/Row/Row.component';
import Column from '../Grid/Column/Column.component';

const getChildren = (children, cols, id) => children.map((child, index) => (
  // eslint-disable-next-line react/no-array-index-key
  <Column col={Math.floor(12 / cols)} key={`card-group-${id}-${index}`}>
    {child}
  </Column>
));

const CardGroup = ({ cols, children, id }) => (
  <Container>
    <Row>
      {getChildren(children, cols, id)}
    </Row>
  </Container>
);

CardGroup.propTypes = {
  id: PropTypes.string.isRequired,
  cols: PropTypes.oneOf([1, 2, 3, 4, 6]),
  children: PropTypes.node,
};

CardGroup.defaultProps = {
  cols: 1,
  children: [],
};

export default CardGroup;
