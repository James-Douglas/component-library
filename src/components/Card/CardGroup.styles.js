import styled from 'styled-components';

export const StyledCardGroup = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: flex-start;
  width: 100%;
  margin: ${({ theme }) => `0 -${theme.spacing['4']}`};
`;

export const StyledCardGroupChildren = styled.div`
  display: flex;
  width: ${({ cols }) => (cols === 1 ? '100%' : `${100 / cols}%`)};
  padding-right: ${({ theme }) => theme.spacing['8']};
  padding-left: ${({ theme }) => theme.spacing['8']};
`;
