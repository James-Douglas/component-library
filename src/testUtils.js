// test-utils.js
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './providers/ManorProvider';

const customRender = (ui, options) => render(ui, { wrapper: ManorProvider, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
