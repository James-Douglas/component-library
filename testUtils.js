import React from 'react';
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './packages/Provider/ManorProvider';

const ManorTestWrapper = ({ children }) => (
  <ManorProvider>
    {children}
  </ManorProvider>
);

const customRender = (ui, options) => render(ui, { wrapper: ManorTestWrapper, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
