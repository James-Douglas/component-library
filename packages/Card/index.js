import Card from './Card.component';
import CardGroup from './CardGroup.component';

export { Card, CardGroup };
