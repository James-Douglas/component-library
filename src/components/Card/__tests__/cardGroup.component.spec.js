import React from 'react';
import { render } from '../../../testUtils';
import 'jest-styled-components';
import Card from '../Card.component';
import CardGroup from '../CardGroup.component';

describe('CardGroup', () => {
  it('renders with minimal props', () => {
    const { container } = render(<CardGroup id="test" />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders with cards in 1 col layout', () => {
    const { container } = render(
      <CardGroup id="test">
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card = container.querySelectorAll('[cols="1"]');
    expect(card[0]).toHaveStyle('width: 100%');
    expect(card[1]).toHaveStyle('width: 100%');
  });

  it('renders with cards in 2 col layout', () => {
    const { container } = render(
      <CardGroup id="test" cols={2}>
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card = container.querySelectorAll('[cols="2"]');
    expect(card[0]).toHaveStyle('width: 50%');
    expect(card[1]).toHaveStyle('width: 50%');
  });

  it('renders with cards in 3 col layout', () => {
    const { container } = render(
      <CardGroup id="test" cols={3}>
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card = container.querySelectorAll('[cols="3"]');
    expect(card[0]).toHaveStyle('width: 33.333333333333336%');
    expect(card[1]).toHaveStyle('width: 33.333333333333336%');
  });

  it('renders with cards in 4 col layout', () => {
    const { container } = render(
      <CardGroup id="test" cols={4}>
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card = container.querySelectorAll('[cols="4"]');
    expect(card[0]).toHaveStyle('width: 25%');
    expect(card[1]).toHaveStyle('width: 25%');
  });

  it('renders with cards in 6 col layout', () => {
    const { container } = render(
      <CardGroup id="test" cols={6}>
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card = container.querySelectorAll('[cols="6"]');
    expect(card[0]).toHaveStyle('width: 16.666666666666668%');
    expect(card[1]).toHaveStyle('width: 16.666666666666668%');
  });

  it('renders with cards in 7 col layout', () => {
    const { container } = render(
      <CardGroup id="test" cols={7}>
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card = container.querySelectorAll('[cols="7"]');
    expect(card[0]).toHaveStyle('width: 14.285714285714286%');
    expect(card[1]).toHaveStyle('width: 14.285714285714286%');
  });

  it('renders with cards in 8 col layout', () => {
    const { container } = render(
      <CardGroup id="test" cols={8}>
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card = container.querySelectorAll('[cols="8"]');
    expect(card[0]).toHaveStyle('width: 12.5%');
    expect(card[1]).toHaveStyle('width: 12.5%');
  });

  it('renders with cards in 9 col layout', () => {
    const { container } = render(
      <CardGroup id="test" cols={9}>
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card = container.querySelectorAll('[cols="9"]');
    expect(card[0]).toHaveStyle('width: 11.11111111111111%');
    expect(card[1]).toHaveStyle('width: 11.11111111111111%');
  });

  it('renders with cards in 10 col layout', () => {
    const { container } = render(
      <CardGroup id="test" cols={10}>
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card = container.querySelectorAll('[cols="10"]');
    expect(card[0]).toHaveStyle('width: 10%');
    expect(card[1]).toHaveStyle('width: 10%');
  });

  it('renders with cards in 11 col layout', () => {
    const { container } = render(
      <CardGroup id="test" cols={11}>
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card = container.querySelectorAll('[cols="11"]');
    expect(card[0]).toHaveStyle('width: 9.090909090909092%');
    expect(card[1]).toHaveStyle('width: 9.090909090909092%');
  });

  it('renders with cards in 12 col layout', () => {
    const { container } = render(
      <CardGroup id="test" cols={12}>
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card = container.querySelectorAll('[cols="12"]');
    expect(card[0]).toHaveStyle('width: 8.333333333333334%');
    expect(card[1]).toHaveStyle('width: 8.333333333333334%');
  });

  it('renders with cards that have margin', () => {
    const margin = ['12', '4', '4', '12'];
    const { container } = render(
      <CardGroup id="test" cols={12} cardProps={{ margin }}>
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card1 = container.querySelector('[id="c1"]');
    const card2 = container.querySelector('[id="c2"]');
    expect(card1).toHaveStyle('margin: 1.2rem 0.4rem 0.4rem 1.2rem');
    expect(card2).toHaveStyle('margin: 1.2rem 0.4rem 0.4rem 1.2rem');
  });

  it('renders with cards that have padding', () => {
    const padding = ['0', '4', '8', '12'];
    const { container } = render(
      <CardGroup id="test" cols={12} cardProps={{ padding }}>
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card1 = container.querySelector('[id="c1"]');
    const card2 = container.querySelector('[id="c2"]');
    expect(card1).toHaveStyle('padding: 0px 0.4rem 0.8rem 1.2rem');
    expect(card2).toHaveStyle('padding: 0px 0.4rem 0.8rem 1.2rem');
  });

  it('renders with cards that have margin and padding', () => {
    const padding = ['4', '4', '8', '12'];
    const margin = ['8', '8', '16', '24'];
    const { container } = render(
      <CardGroup id="test" cols={12} cardProps={{ padding, margin }}>
        <Card id="c1" />
        <Card id="c2" />
      </CardGroup>,
    );
    const card1 = container.querySelector('[id="c1"]');
    const card2 = container.querySelector('[id="c2"]');
    expect(card1).toHaveStyle('padding: 0.4rem 0.4rem 0.8rem 1.2rem');
    expect(card1).toHaveStyle('margin: 0.8rem 0.8rem 1.6rem 2.4rem');
    expect(card2).toHaveStyle('padding: 0.4rem 0.4rem 0.8rem 1.2rem');
    expect(card2).toHaveStyle('margin: 0.8rem 0.8rem 1.6rem 2.4rem');
  });
});
