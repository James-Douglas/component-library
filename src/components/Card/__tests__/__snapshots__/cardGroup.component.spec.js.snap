// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardGroup renders with minimal props 1`] = `"<div id=\\"test\\" class=\\"sc-bwzfXH eXuFVT\\"></div>"`;
